### PR TITLE
API: no_default

### LINKED ISSUE #30785: Rename api.extensions._no_default to extension.no_default
Follow-up on https://github.com/pandas-dev/pandas/pull/30322, which exposed `lib._no_default` in `pandas.api.extensions`

### PR DESCRIPTION
**Pull request by TomAugspurger**
Changes lib._no_default to lib.no_default, uses it in more places.

Closes #30785

### CODE REVIEW

**Comment by WillAyd** [comment]
lgtm if green

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2270,7 +2270,7 @@ def map_infer_mask(ndarray arr, object f, const uint8_t[:] mask, bint convert=1,
     result = np.empty(n, dtype=dtype)
     for i in range(n):
         if mask[i]:
-            if na_value is _no_default:
+            if na_value is no_default:
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx]: having a cdef version (kind of like we have for NaT) would help perf here (not necessary for this PR)

File: pandas/api/extensions/__init__.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 """Public API for extending pandas objects."""
-from pandas._libs.lib import _no_default  # noqa: F401
+from pandas._libs.lib import no_default  # noqa: F401
```
Comment (jbrockmendel) [inline pandas/api/extensions/__init__.py]: we want this available to library authors, but we never want users to use this, right?

Comment (TomAugspurger) [reply pandas/api/extensions/__init__.py]: It should be fine for users. I can imagine people working with the signature of functions that use no_default.

**Comment by TomAugspurger** [comment]
Merging since I have a followup depending on this. Will respond more on https://github.com/pandas-dev/pandas/pull/30788#discussion_r363858069 if needed.

**Pull request merged**